irobot_events_executor/include/rclcpp/executors/events_executor/events_executor_notify_waitable.hpp

```diff
@@ -29,7 +29,14 @@ class EventsExecutorNotifyWaitable final : public EventWaitable
 
   // Destructor
   RCLCPP_PUBLIC
-  virtual ~EventsExecutorNotifyWaitable() = default;
+  virtual ~EventsExecutorNotifyWaitable()
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
```

Comment:
I'm not convinced by this mutex.
Even with it, people should not interact with the waitable while a thread is destroying it

Reply:
The same situation applies to the simple events queue destructor; I'm not sure if the fix is correct.

Reply:
I can remove the mutex, I thought it was the cause of a segfault I was having but wasn't.

But the callbacks should be unset anyway, there's a time gap in which the guard conditions can be triggered and will segfault. This is the ordering of destruction of things:
```
~EventsExecutor
~EventsExecutorEntitiesCollector

// From here one, triggering these gc's will segfault
~Executor: ~interrupt_guard_condition_
~Executor: ~shutdown_guard_condition_
```

Reply:
I agree on unsetting the callbacks